model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java

```diff
@@ -338,7 +338,8 @@ public boolean removeRole(RoleModel role) {
         }
         String compositeRoleTable = JpaUtils.getTableNameForNativeQuery("COMPOSITE_ROLE", em);
         em.createNativeQuery("delete from " + compositeRoleTable + " where CHILD_ROLE = :role").setParameter("role", roleEntity).executeUpdate();
-        realm.getClientsStream().forEach(c -> c.deleteScopeMapping(role));
+        String clientScopeMapping = JpaUtils.getTableNameForNativeQuery("SCOPE_MAPPING", em);
```

Comment:
The method `removeRole` is part of the `RoleProvider` and as such cannot assume that the client scope mappings would be stored in the same JPA store - this holds for the legacy JPA store but will not necessarily be true for new storage.

The right approach would be to introduce a method to delete a client scope mappings based on the role. @vramik, can you please help here?

Reply:
Actually, looking a bit further into the code, this should be introduced as a method `preRemove` in a `ClientProvider` that would take a parameter `realm` and `role`, similarly to what's done on L 344 for groups.

Reply:
Thanks @hmlnarik! I think that I've understood what you mean. I'll add the `preRemove` to the `ClientProvider` method and update the necessary implementations. I'll try to submit a new PR next Monday.

Reply:
Thank you @rmartinc for the PR.

+1 for introduction of `preRemove` method.

Reply:
OK, it's done. There are two things to comment:

1. The method cannot be named `preRemove` because there are two classes that implements Group and Client provider and if the methods have the same signature, the same method is executed several times (x4, x2 in `JpaRealmProvider` and another x2 in `RealmCacheSession`). So I have used the name `preRemoveClientProvider` for the client provider.
2. In the `MapClientProvider` no new searchable field was created, all the clients in the realm are iterated to update them. I think this part is WIP so I prefer to modify it as little as possible.

Just let me know if you want more modifications. Thanks!